File: 01_Arrays_Hashing/Easy/python/valid_anagram.py
"""
-----Approach 1 (Optimal solution)-------

-> Both strings contain only lowercase English letters (a-z).
-> Inputs can be empty.
-> Use an array of size 26 to count character frequencies.

Algorithm:
    -> If lengths of s and t differ, return False.
    -> Count the frequency of each character in s.
    -> Decrease the frequency for each character in t.
    -> If all counts return to zero, return True.

Complexity:
    -> Time Complexity: O(n)
    -> Space Complexity: O(1)
"""


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t): 
            return False
        count = [0] * 26
        for c in s:
            count[ord(c) - ord('a')] += 1

        for c in t:
            count[ord(c) - ord('a')] -= 1
            if count[ord(c) - ord('a')] < 0:
                return False
        return True

File: 01_Arrays_Hashing/Easy/c#/ValidAnagram.cs
// Problem: Given two strings s and t, return true if t is an anagram of s, and false otherwise.
//
// --- Approach ---
// Use two dictionaries (hash maps) to count the frequency of each character in both strings.
// Then compare the counts: if they match for all characters, the strings are anagrams.
//
// Steps:
// 1-Check if lengths are different → return false immediately.
// 2-Count character occurrences for each string using Dictionary<char, int>.
// 3-Compare frequency tables of both dictionaries.
// 4-Return true only if all character counts match.
//
// Time Complexity: O(n)
// Space Complexity: O(1) → technically O(26) for lowercase English letters, or O(k) for unique characters.

public class Solution {
    public bool IsAnagram(string s, string t)
    {
        if (s.Length != t.Length)
            return false;

        Dictionary<char, int> ds = new Dictionary<char, int>();
        Dictionary<char, int> dt = new Dictionary<char, int>();

        foreach (var ch in s)
        {
            if (ds.ContainsKey(ch))
                ds[ch]++;
            else
                ds[ch] = 1;
        }

        foreach (var ch in t)
        {
            if (dt.ContainsKey(ch))
                dt[ch]++;
            else
                dt[ch] = 1;
        }

        foreach (var kv in ds)
        {
            if (!dt.ContainsKey(kv.Key) || dt[kv.Key] != kv.Value)
                return false;
        }

        return true;
    }

}
